src/lib.rs

```diff
@@ -0,0 +1,389 @@
+//! use fm::FMatcher;
+//!
+//! assert!(FMatcher::new("a").unwrap().matches("a").is_ok());
+//! assert!(FMatcher::new(" a ").unwrap().matches("a").is_ok());
+//! assert!(FMatcher::new("a").unwrap().matches("b").is_err());
+//! assert!(FMatcher::new("a\n...\nb").unwrap().matches("a\na\nb").is_ok());
+//! assert!(FMatcher::new("a\n...\nb").unwrap().matches("a\na\nb\nb").is_err());
+//! ```
+//!
+//! If you want to use non-default options, you will first need to use `FMBuilder` before having
+//! access to an `FMatcher`.
+
+use std::{
+    collections::hash_map::{Entry, HashMap},
+    error::Error,
+};
+
+use regex::Regex;
+
+const WILDCARD: &str = "...";
+
+struct FMOptions {
+    name_matcher: Option<(Regex, Regex)>,
+    ignore_leading_whitespace: bool,
+    ignore_trailing_whitespace: bool,
+}
+
+/// Build up a `FMatcher` allowing the setting of options.
+///
+/// ```rust
+/// use {fm::FMBuilder, regex::Regex};
+///
+/// let ptn_re = Regex::new("\\$.+?\\b").unwrap();
+/// let text_re = Regex::new(".+?\\b").unwrap();
+/// let matcher = FMBuilder::new("$1 $1")
+///                         .unwrap()
+///                         .name_matcher(Some((ptn_re, text_re)))
+///                         .build()
+///                         .unwrap();
+/// assert!(matcher.matches("a a").is_ok());
+/// assert!(matcher.matches("a b").is_err());
+/// ```
+pub struct FMBuilder<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMBuilder<'a> {
+    /// Create a new `FMBuilder` with default options.
+    pub fn new(ptn: &'a str) -> Result<Self, Box<dyn Error>> {
+        let options = FMOptions {
+            name_matcher: None,
+            ignore_leading_whitespace: true,
+            ignore_trailing_whitespace: true,
+        };
+        Ok(FMBuilder { ptn, options })
+    }
+
+    /// Add a name matcher `Some((ptn_re, text_re))` (or unset it with `None`). Defaults to `None`.
+    ///
+    /// Name matchers allow you to ensure that different parts of the text match without specifying
+    /// precisely what they match. For example, if you have output where you want to ensure that
+    /// two locations always match the same name, but the name is non-deterministic you can allow
+    /// the use of `$` wildcards in your pattern:
+    ///
+    /// ```rust
+    /// use {fm::FMBuilder, regex::Regex};
+    ///
+    /// let ptn_re = Regex::new("\\$.+?\\b").unwrap();
+    /// let text_re = Regex::new(".+?\\b").unwrap();
+    /// let matcher = FMBuilder::new("$1 b $1")
+    ///                         .unwrap()
+    ///                         .name_matcher(Some((ptn_re, text_re)))
+    ///                         .build()
+    ///                         .unwrap();
+    /// assert!(matcher.matches("a b a").is_ok());
+    /// assert!(matcher.matches("a b b").is_err());
+    /// ```
+    pub fn name_matcher(mut self, matcher: Option<(Regex, Regex)>) -> Self {
+        self.options.name_matcher = matcher;
+        self
+    }
+
+    #[doc(hidden)]
```

Comment:
How comes?

Reply:
Because if you set them to `false` you hit a `todo` :) It just felt like too much for one PR.

Reply:
Ah yes. OK. You could even un-pub them, or remove them from this PR. Up to you.

Reply:
I think we can just leave them in and flesh them out later.